ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/GSM/Model.hs

```diff
@@ -0,0 +1,652 @@
+        Unit
+    TimePasses{} ->
+        Unit
+  where
+    j = toJudgment $ mState model
+
+-----
+
+-- | A block count
+newtype B = B Int
+  deriving stock    (Eq, Ord, Generic, Read, Show)
+  deriving newtype  (Enum, Num)
+  deriving anyclass (TD.ToExpr)
+
+-- | A slot count
+newtype S = S Int
+  deriving stock    (Eq, Ord, Generic, Read, Show)
+  deriving newtype  (Enum, Num)
+  deriving anyclass (TD.ToExpr)
+
+data UpstreamPeer = Amara | Bao | Cait | Dhani | Eric
+  deriving stock    (Bounded, Enum, Eq, Ord, Generic, Read, Show)
+  deriving anyclass (TD.ToExpr)
+
+-- | The cumulative growth relative to whatever length the initial selection
+-- was and the slot relative to the start of the test (which is assumed to be
+-- the exact onset of some slot)
+data Selection = Selection !B !S
+  deriving stock    (Eq, Ord, Generic, Show)
+  deriving anyclass (TD.ToExpr)
+
+-- | The age of the candidate is irrelevant, only its length matters
+newtype Candidate = Candidate B
+  deriving stock    (Eq, Ord, Generic, Show)
+  deriving anyclass (TD.ToExpr)
+
+data MarkerState = Present | Absent
+  deriving stock    (Eq, Ord, Generic, Read, Show)
+  deriving anyclass (TD.ToExpr)
+
+newtype WhetherPrevTimePasses = WhetherPrevTimePasses Bool
+  deriving stock    (Eq, Ord, Generic, Show)
+  deriving anyclass (TD.ToExpr)
+
+data ModelState =
+    ModelTooOld
+  |
+    ModelYoungEnough !SI.Time
+    -- ^ when the model most recently transitioned to 'YoungEnough'
+  deriving stock    (Eq, Ord, Generic, Show)
+  deriving anyclass (TD.ToExpr)
+
+-----
+
+-- | Interesting events to record /within the model/
+--
+-- TODO some less superficial ones (eg even just combinations of these)
+data Notable =
+    BigDurN
+    -- ^ there was a "big" 'TimesPasses' command
+  |
+    CaughtUpN
+    -- ^ the node transitioned from OnlyBootstrap to CaughtUp
+  |
+    FellBehindN
+    -- ^ the node transitioned from CaughtUp to OnlyBootstrap
+  |
+    FlickerN
+    -- ^ the node transitioned from CaughtUp to OnlyBootstrap and back to
+    -- CaughtUp "instantly"
+  |
+    NotThrashingN
+    -- ^ the anti-thrashing would have allowed 'FellBehindN', but the selection
+    -- wasn't old enough
+  |
+    TooOldN
+    -- ^ the selection was old enough for 'FellBehindN', but the anti-thrashing
+    -- prevented it
+  deriving (Eq, Ord, Show)
+
+instance TD.ToExpr Notable where toExpr = TD.defaultExprViaShow
+
+----- orphans
+
+instance TD.ToExpr SI.Time              where toExpr = TD.defaultExprViaShow
+instance TD.ToExpr LedgerStateJudgement where toExpr = TD.defaultExprViaShow
+
+deriving instance Read LedgerStateJudgement
+
+instance QC.Arbitrary LedgerStateJudgement where
+    arbitrary = elements [TooOld, YoungEnough]
+    shrink    = \case
+        TooOld      -> [YoungEnough]
+        YoungEnough -> []
+
+instance QC.Arbitrary MarkerState where
+    arbitrary = elements [Absent, Present]
+    shrink    = \case
+        Absent  -> [Present]
+        Present -> []
```

Comment:
These could be moved to `Test.Util.Orphans.{ToExpr,Arbitrary}` (or maybe even upstreamed for `LedgerStateJudgement` @bolt12)?

Reply:
To clarify: `MarkerState ` is not an orphan instance.

But the three `LedgerStateJudgement` instances are. And the `SI.Time` as well.

Reply:
`LedgerStateJudgement` already has an arbitrary instance but it is given by `ArbitraryLedgerStateJudgement`, however I am not sure if you can access it

Reply:
It should exist in `ouroboros-network:sim-tests-lib`